lib/Feature/FeatureModelParser.cpp

```diff
@@ -225,4 +252,361 @@ std::unique_ptr<xmlDoc, void (*)(xmlDocPtr)> FeatureModelXmlParser::parseDoc() {
+
+std::unique_ptr<xmlDtd, void (*)(xmlDtdPtr)>
+FeatureModelSxfmParser::createDtd() {
+  std::unique_ptr<xmlDtd, void (*)(xmlDtdPtr)> Dtd(
+      xmlIOParseDTD(nullptr,
+                    xmlParserInputBufferCreateMem(SxfmConstants::DtdRaw.c_str(),
+                                                  SxfmConstants::DtdRaw.length(),
+                                                  XML_CHAR_ENCODING_UTF8),
+                    XML_CHAR_ENCODING_UTF8),
+      xmlFreeDtd);
+  xmlCleanupParser();
+  assert(Dtd && "Failed to parse DTD.");
+  return Dtd;
+}
+
+bool FeatureModelSxfmParser::parseFeatureTree(xmlChar *FeatureTree) {
+  // Split the lines of the feature tree by new lines
+  {
+    std::stringstream Ss(reinterpret_cast<const char *>(FeatureTree));
+    std::string To;
+    string Name;
+    bool Opt;
+    int LastIndentationLevel = -1;
+    int RootIndentation = -1;
+    int OrGroupCounter = 0;
+    std::map<int, string> IndentationToParentMapping;
+
+    // This map is used for the or group mapping
+    // Each entry represents an or group as a tuple where the first value is
+    // the name of the parent, the second is the relationship kind, and the
+    // third a vector consisting of the name of the children
+    std::map<int, std::tuple<string, Relationship::RelationshipKind, std::vector<string>>> OrGroupMapping;
+
+    if (FeatureTree == nullptr) {
+      std::cerr << "Failed to read in feature tree. Is it empty?" << std::endl;
+      return false;
+    }
+
+    while (std::getline(Ss, To)) {
+      Opt = false;
+
+      if (To.empty() || std::all_of(To.begin(),To.end(),isspace)) {
+        continue;
+      }
+
+      // For every line, count the indentation
+      // not more than 1 additional indentations are allowed to the original one
+      // However, we may have arbitrarily less indentations
+      if ((To.find(':', 0 )) == std::string::npos) {
+        std::cerr << "Colon is missing in line" << To << std::endl;
+        return false;
+      }
+
+      int CurrentIndentationLevel = countOccurrences(readUntil(To, ':', 0), Indentation);
+      int Diff = CurrentIndentationLevel - LastIndentationLevel;
+
+      // Remember the root indentation for later checks
+      if (LastIndentationLevel == -1) {
+        RootIndentation = CurrentIndentationLevel;
+      }
+
+      if ((LastIndentationLevel != -1) && Diff > 1) {
+        std::cerr << "Indentation error in feature tree in line " << To << std::endl;
+        return false;
+      }
+
+      // Move pointer to first character after indentation
+      // The first character has to be a colon followed by the type of
+      // the feature (m for mandatory, o for optional, a for alternative)
+      std::string::size_type Pos = CurrentIndentationLevel * Indentation.length() + 2;
+      std::optional<std::tuple<int, int>> Cardinalities;
+
+      switch (To.at(Pos - 1)) {
+      case 'o':
+        // Code for optional
```

Comment:
If these code have special meaning, would it make sense to create a small enum that make the mapping clear? So we could write:
```
case SXFM_CHAR_CODE::OPTIONAL:
    // ... do stuff
```